pkg/client/block/client.go

```diff
@@ -36,42 +39,147 @@ func NewBlockClient(
 	ctx context.Context,
 	deps depinject.Config,
 ) (client.BlockClient, error) {
-	client, err := events.NewEventsReplayClient[client.Block](
+	ctx, close := context.WithCancel(ctx)
+
+	eventsReplayClient, err := events.NewEventsReplayClient[client.Block](
 		ctx,
 		deps,
 		committedBlocksQuery,
 		newCometBlockEvent,
 		defaultBlocksReplayLimit,
 	)
 	if err != nil {
+		close()
+		return nil, err
+	}
+
+	// latestBlockPublishCh is the channel that notifies the latestBlockReplayObs of a
+	// new block, whether it comes from a direct query or an event subscription query.
+	latestBlockReplayObs, latestBlockPublishCh := channel.NewReplayObservable[client.Block](ctx, defaultBlocksReplayLimit)
+	blockClient := &blockReplayClient{
+		eventsReplayClient:   eventsReplayClient,
+		latestBlockReplayObs: latestBlockReplayObs,
+		close:                close,
+	}
+
+	if err := depinject.Inject(deps, &blockClient.onStartQueryClient); err != nil {
 		return nil, err
 	}
-	return &blockClient{eventsReplayClient: client}, nil
+
+	blockClient.forEachBlockEvent(ctx, latestBlockPublishCh)
+
+	if err := blockClient.getInitialBlock(ctx, latestBlockPublishCh); err != nil {
+		return nil, err
+	}
+
+	return blockClient, nil
```

Comment:
```suggestion
	return blockClient.getInitialBlock(ctx, latestBlockPublishCh)
```

Reply:
We need to return a `blockClient, err` tuple